File: linked_list/LL_heap_pq/maxheap_LL.py
class Node:
    """A Node of the linked list."""
    def __init__(self, val):
        self.val = val
        self.next = None


class MaxHeap:
    """Max-Heap implemented as a linked list."""
    def __init__(self):
        self.head = None

    def insert(self, val):
        """Insert a value into the Max-Heap."""
        new_node = Node(val)
        if not self.head or val > self.head.val:
            new_node.next = self.head
            self.head = new_node
        else:
            current = self.head
            while current.next and current.next.val >= val:
                current = current.next
            new_node.next = current.next
            current.next = new_node

    def pop(self):
        """Remove and return the largest value from the Max-Heap."""
        if not self.head:
            raise IndexError("Max-Heap is empty!")
        max_val = self.head.val
        self.head = self.head.next
        return max_val

    def sort(self):
        """
        Sort the Max-Heap in ascending order.
        - Use a selection sort-like approach directly on the linked list.
        """
        sorted_head = None  # Head of the sorted list

        while self.head:
            # Find the maximum node in the current heap
            max_node = self.head
            max_prev = None
            current = self.head
            prev = None

            while current and current.next:
                if current.next.val > max_node.val:
                    max_node = current.next
                    max_prev = current
                current = current.next

            # Remove the max_node from the original list
            if max_prev:
                max_prev.next = max_node.next
            else:
                self.head = max_node.next

            # Insert max_node at the beginning of the sorted list
            max_node.next = sorted_head
            sorted_head = max_node

        self.head = sorted_head

    def print_heap(self):
        """Print the Max-Heap."""
        current = self.head
        while current:
            print(current.val, end=" ")
            current = current.next
        print()


# Test Max-Heap with sort functionality
print("\nMax-Heap with Linked List Implementation:")
max_heap = MaxHeap()

max_heap.insert(10)
max_heap.insert(20)
max_heap.insert(5)
max_heap.insert(15)

print("Max-Heap after insertions: ", end="")
max_heap.print_heap()

# Sort the Max-Heap
print("\nSorting the Max-Heap:")
max_heap.sort()
max_heap.print_heap()

File: linked_list/LL_heap_pq/maxheap_LL.cpp
#include <iostream>
#include <vector>
using namespace std;

/*
 * Node 구조체: 연결 리스트의 각 노드를 정의.
 * MaxHeap 클래스: 연결 리스트를 기반으로 한 최대 힙 구현.
 *
 * 주요 기능:
 * - Node: 정수 값을 저장하며, 다음 노드 포인터를 포함.
 * - MaxHeap:
 *   1. insert: 새로운 값을 최대 힙에 삽입.
 *   2. pop: 최대값 제거 및 반환.
 *   3. print: 현재 힙 상태 출력.
 *   4. sort: 힙 요소를 내림차순으로 정렬 (추가 구현).
 */

// Node 구조체 정의
struct Node {
    int val;       // 노드 값
    Node* next;    // 다음 노드 포인터

    // 생성자
    Node(int v) : val(v), next(nullptr) {}
};

// MaxHeap 클래스 정의
class MaxHeap {
private:
    Node* head; // 연결 리스트의 시작 노드

    // 새로운 노드를 생성하는 헬퍼 함수
    Node* createNode(int value) {
        return new Node(value);
    }

    // 연결 리스트를 배열로 변환 (정렬에 필요)
    vector<int> toArray() const {
        vector<int> array;
        Node* current = head;
        while (current) {
            array.push_back(current->val);
            current = current->next;
        }
        return array;
    }

    // 배열을 연결 리스트로 변환 (정렬 후 재구성)
    void fromArray(const vector<int>& array) {
        clear(); // 기존 리스트 제거
        for (int value : array) {
            insert(value);
        }
    }

    // 연결 리스트 초기화 (메모리 해제)
    void clear() {
        while (head) {
            Node* temp = head;
            head = head->next;
            delete temp;
        }
    }

public:
    // 생성자: head 초기화
    MaxHeap() : head(nullptr) {}

    // 최대 힙에 값 삽입
    void insert(int value) {
        Node* newNode = createNode(value);
        if (!head || newNode->val > head->val) {
            newNode->next = head;
            head = newNode;
        } else {
            Node* current = head;
            while (current->next && current->next->val >= newNode->val) {
                current = current->next;
            }
            newNode->next = current->next;
            current->next = newNode;
        }
    }

    // 최대값 제거 및 반환
    int pop() {
        if (!head) {
            throw runtime_error("Max-Heap is empty!");
        }
        int maxValue = head->val;
        Node* temp = head;
        head = head->next;
        delete temp;
        return maxValue;
    }

    // 힙 출력
    void print() const {
        Node* current = head;
        while (current) {
            cout << current->val << " ";
            current = current->next;
        }
        cout << endl;
    }

    // 힙 정렬 (내림차순)
    void sort() {
        vector<int> array = toArray(); // 힙을 배열로 변환
        // 배열을 내림차순 정렬
        std::sort(array.begin(), array.end(), greater<int>());
        // 정렬된 배열을 연결 리스트로 변환
        fromArray(array);
    }

    // 소멸자: 메모리 해제
    ~MaxHeap() {
        clear();
    }
};

// 메인 함수: MaxHeap 테스트
int main() {
    cout << "\nMax-Heap with Linked List Implementation:\n";

    MaxHeap heap;

    // 삽입 테스트
    heap.insert(10);
    heap.insert(20);
    heap.insert(5);
    heap.insert(15);

    // 힙 출력
    cout << "Max-Heap after insertions: ";
    heap.print();

    // 정렬 테스트
    cout << "\nSorting Max-Heap (Descending Order):\n";
    heap.sort();
    heap.print();

    // 최대값 제거 테스트
    cout << "\nPopping maximum: " << heap.pop() << endl;
    cout << "Max-Heap after popping: ";
    heap.print();

    return 0;
}